### ISSUE TITLE
Improve performance for accessing files from IDOM_WEB_MODULES_DIR

### ISSUE THREAD

**Issue by rmorshea** [idom-team/django-idom#4]
`IDOM_WEB_MODULES_DIR` contains JS modules that are dynamically added at runtime. Since JS files are typically loaded statically, it would be nice if we could improve how quickly these files can be accessed. There are three ways to do that:

- [ ] Implement a management command for collecting these static files ahead of time.
- [x] Make the view accessing the files `async` to avoid blocking while waiting on the file system
- [x] Cache this view using [Django's cache framework](https://docs.djangoproject.com/en/3.2/topics/cache/).

### Notes

To implement the management command we could trace over all the templates and look for `idom_component` template tag nodes in order to figure out what modules to import so that we can trigger the collection of the JS files into `IDOM_WEB_MODULES_DIR`. Then once they've been collected those files can be copied into a permanent static location.

To do this we can take inspiration from the [django compressor](https://github.com/django-compressor/django-compressor) library which takes a similar approach. See the `COMPRESS_OFFLINE` setting to figure out how they're doing it. Also the `compressor.offiline` module contains uniform interfaces for parsing template syntax for Django and Jinja that we might want to copy (with the latest [LICENSE](https://github.com/django-compressor/django-compressor/blob/develop/LICENSE), what to do about `AUTHORS` file?).

---

_Conversation originated from: [this comment](https://github.com/idom-team/django-idom/pull/3/files/4bcdeb522ad86d46ad7e3bee13ad51bdc32026fc#r688325126)_

**Comment by rmorshea** [comment]
@Archmonger I'm looking at what's required to implement this and its kind of ugly stuff. I'd mostly be copy-pasting code I don't really understand which makes me nervous about maintaining it going forward. I'm also still a bit skeptical that this will actually be a bottle neck. Given that, I think I'm going to hold off for now. If it does prove to be a real issue, we can revisit at that point.

With that said, I think there are two simpler solutions to this problem of the dynamic files in `IDOM_WEB_MODULES_DIR` that can improve performance and will be easier to maintain:

1. Make the view reading these web modules `async` so it can `await` the file instead of blocking until the file system is available.
2. Allow users to configure a cross-process cache using [Django's cache framework](https://docs.djangoproject.com/en/3.2/topics/cache/).

**Comment by Archmonger** [comment]
I agree that it should be async. Generally everything that supports Django 3.0+ only should be async, unless you're DB reliant (Django team is still working on async DB ORM).

I also was having trouble following their code, but I'll see if I can digest their code and make it maintainable.

If not, I do have another idea in mind to accomplish the same thing.

We could directly call `django.contrib.staticfiles.finders.AppDirectoriesFinder` and, as you suggested before, do a regex match on all `.html` files for the idom component tag.

**Comment by rmorshea** [comment]
> I'm also still a bit skeptical that this will actually be a bottle neck.

The _this_ in that statement referred to accessing the JS files. That is, that accessing the JS files using the "dynamic" approach (as apposed to being able to load them from a static dir) would become a bottleneck.

For now, I think I'm going to make the function async, and add the basic caching with `functools` we'd already talked about. I'll leave this issue open, and mark with with the "future" label.

Since a static collection method would be a backwards compatible addition, I think I feel comfortable releasing with the aforementioned changes so I can move on to other things in `idom` core.

**Comment by Archmonger** [comment]
If we do intend to implement this as a templatetag later, I've determined the relevant lines of code related to django_compressor. Although their code is severely lacking comments, I personally think it's fairly easy to follow.

See [`django-compressor/compress.py:80-233`](https://github.com/django-compressor/django-compressor/blob/f533dc6dde9ed90626382d78f3eb37b84d848027/compressor/management/commands/compress.py#L90).

### Psuedocode

1) [Create a list of all templates via `get_loaders()`.](https://github.com/django-compressor/django-compressor/blob/f533dc6dde9ed90626382d78f3eb37b84d848027/compressor/management/commands/compress.py#L104)
2) [Check all templates for broken syntax via either jinja2 `parse()` or django `parse()`. Collect all non-broken templates into a list.](https://github.com/django-compressor/django-compressor/blob/f533dc6dde9ed90626382d78f3eb37b84d848027/compressor/management/commands/compress.py#L160)
3) [Perform `_compress_template` on each template tag node via a `ThreadPoolExecutor`.](https://github.com/django-compressor/django-compressor/blob/f533dc6dde9ed90626382d78f3eb37b84d848027/compressor/management/commands/compress.py#L192)
   - This notably uses [`django-compressor.offline.django.walk_nodes`](https://github.com/django-compressor/django-compressor/blob/f533dc6dde9ed90626382d78f3eb37b84d848027/compressor/offline/django.py#L141) and [`django-compressor.offline.jinja2.walk_nodes`](https://github.com/django-compressor/django-compressor/blob/f533dc6dde9ed90626382d78f3eb37b84d848027/compressor/offline/jinja2.py#L118) to collect the templatetags on the page.
   - By default it looks like they're collecting this info using Python's built-in `HTMLParser`, but it's extensible to utilize `lxml`, `beautifulsoup`, or `html5lib`. We could _probably_ replace this with regex.

### Future Implementation

Since we're leaving this for later, we're going to need to follow in django-compressor's footsteps of having a `COLLECT_OFFLINE` setting for backwards compatibly.

**Comment by Archmonger** [comment]
On the note of performance, we may also want to consider caching the output of [`_register_component`](https://github.com/idom-team/django-idom/blob/90db8d93b8368afd21408bd933680822a4f8939c/src/django_idom/templatetags/idom.py#L34)

**Comment by rmorshea** [comment]
[This check](https://github.com/idom-team/django-idom/blob/90db8d93b8368afd21408bd933680822a4f8939c/src/django_idom/templatetags/idom.py#L37-L38) should cover that since we avoid attempting an import if it's already in the set of imported modules.

**Comment by rmorshea** [comment]
Ah, but one could use the same component more than once, so we could skip registration entirely if it's already there.

**Comment by Archmonger** [comment]
Caching the register function with `funcutils.cache` would provide the same ability to skip registration, and would allow us to skip the `sys.modules` check.

What do you think?

**Comment by Archmonger** [comment]
### Progress Update
A lot of this issue has been resolved. Here's a list of a few things that have already been developed

- [x] Async views
- [x] Web module caching
- [x] Component tag parser inspired by django-compressor

The following task is remaining

- [ ] Management command for collecting static files ahead of time

**Comment by Archmonger** [comment]
I'm going to close this issue, as the fully async web module fetching we have developed is good enough for now.

Will create a separate issue specific to `COLLECT` management command.
